File: src/scaling.jl
function resample_df(df, stochastic=true)
    n = nrow(df)
    if stochastic
        ii = sample(1:n, n)
    else
        ii = 1:n
    end
    return @view df[ii, :]
end

function resample_scaling(df, stochastic=true)
    return DataFramesMeta.combine(x -> resample_df(x, stochastic),
        DataFramesMeta.groupby(df, [:haul_id, :class]))
end

function get_trawl_category_means(scaling, aged_species, predict_weight)
    use_ages = in(aged_species)
    trawl_means_cat = @chain scaling begin
        DataFramesMeta.@transform(
            :category = _category.(use_ages, :species_code, :age),
            :weight = predict_weight.(:primary_length),
            :p_nasc = :sigma_bs .* :w
        )
        #=
        If a catch filter is applied, all specimens may be listed twice in the same haul, 
        with "user_defined_expansion" set to 0.0 for one listing and 1.0 for the other.
        This is done to split a haul into two parts, each applied to a different scaling
        stratum. Eliminate weighting factors == 0 here, since they don't affect any of the
        numbers and introduce NaNs if all specimens in a haul/category have :w .== 0.
        =# 
        @subset(:w .> 0)
        @by([:haul_id, :category], 
            :sigma_bs = mean(:sigma_bs),#, Weights(:w)),
            :p_nasc = sum(:p_nasc),
            :weight = mean(:weight)#, Weights(:w)),
        )
        # make proportions sum to 1.0 for each haul
        DataFrames.groupby(:haul_id)
        DataFramesMeta.@transform(:p_nasc = :p_nasc ./ sum(:p_nasc))
    end
    return trawl_means_cat
end

function _category(use_ages, species_code, age)
    if use_ages(species_code)
        return string.(species_code) .* "@" .* string.(age)
    else
        return string.(species_code) .* "@-1"
    end
end

File: python/atbootstrap/scaling.py
"""Scaling utilities."""
from __future__ import annotations

import numpy as np
import pandas as pd


def resample_df(df: pd.DataFrame, stochastic: bool = True) -> pd.DataFrame:
    n = len(df)
    if stochastic:
        ii = np.random.randint(0, n, n)
    else:
        ii = np.arange(n)
    return df.iloc[ii].copy()


def resample_scaling(df: pd.DataFrame, stochastic: bool = True) -> pd.DataFrame:
    return (
        df.groupby(["haul_id", "class"], as_index=False, group_keys=False)
        .apply(lambda x: resample_df(x, stochastic))
        .reset_index(drop=True)
    )


def _category(use_ages, species_code, age):
    if use_ages(species_code):
        return f"{species_code}@{age}"
    return f"{species_code}@-1"


def get_trawl_category_means(scaling: pd.DataFrame, aged_species, predict_weight):
    use_ages = lambda s: s in aged_species
    df = scaling.copy()
    df["category"] = df.apply(lambda r: _category(use_ages, r["species_code"], r["age"]), axis=1)
    df["weight"] = df["primary_length"].apply(predict_weight)
    df["p_nasc"] = df["sigma_bs"] * df["w"]
    df = df[df["w"] > 0]

    trawl_means_cat = (
        df.groupby(["haul_id", "category"], as_index=False)
        .agg(sigma_bs=("sigma_bs", "mean"),
             p_nasc=("p_nasc", "sum"),
             weight=("weight", "mean"))
    )
    trawl_means_cat["p_nasc"] = trawl_means_cat.groupby("haul_id")["p_nasc"].transform(
        lambda x: x / x.sum()
    )
    return trawl_means_cat
